hack/generator/pkg/codegen/pipeline_collapse_cross_group_refs.go

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+)
+
+// collapseCrossGroupReferences finds and removes references between API groups. This isn't particularly common
+// but does occur in a few instances, for example from Microsoft.Compute -> Microsoft.Compute.Extensions.
+func collapseCrossGroupReferences() PipelineStage {
+	return MakePipelineStage(
+		"collapseCrossGroupReferences",
+		"Finds and removes cross group references",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			resources := astmodel.CollectResourceDefinitions(types)
+			result := make(astmodel.Types)
+
+			for resourceName := range resources {
+				walker := newTypeWalker(types, resourceName)
+				updatedTypes, err := walker.Walk(types[resourceName])
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed walking types")
+				}
+
+				for _, newDef := range updatedTypes {
+					err := result.AddAllowDuplicates(newDef)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+
+			return result, nil
+		})
+}
+
+func newTypeWalker(types astmodel.Types, resourceName astmodel.TypeName) *astmodel.TypeWalker {
+	visitor := astmodel.TypeVisitorBuilder{}.Build()
+	walker := astmodel.NewTypeWalker(types, visitor)
+	walker.AfterVisit = func(original astmodel.TypeDefinition, updated astmodel.TypeDefinition, ctx interface{}) (astmodel.TypeDefinition, error) {
+		if !resourceName.PackageReference.Equals(updated.Name().PackageReference) {
+			// Note: If we ever find this generating colliding names, we might need to introduce a unique suffix.
+			// For now though it doesn't seem to, so preserving the shorter names as they're clearer.
+			updated = updated.WithName(astmodel.MakeTypeName(resourceName.PackageReference, updated.Name().Name()))
```

Comment:
Presumably in the case of a collision we'll panic rather than just overwriting?

Reply:
A collision would be `err := result.AddAllowDuplicates(newDef)` above returning an error - meaning we had two types that were named the same but are structurally different. Right now that'd just block the whole pipeline and we'd have to come and add handling to deal with it.